### PR TITLE
Adding 0 and 1 coords if missed in colormap

### LINKED ISSUE #2400: Pad colors when not all control points specified
## To Reproduce

```python
import numpy as np

import napari

colormap = napari.utils.Colormap(colors=[[0,0,0,1], [1, 1, 1, 1]], name="test", controls=[0.1, 0.9] )

print(colormap.colorbar)

with napari.gui_qt():
    viewer = napari.Viewer()
    viewer.add_image(np.random.random(size=(100, 100)), colormap=colormap)
```

end of stacktrace
```
~/.pyenv/versions/3.8.3/envs/napari/lib/python3.8/site-packages/vispy/color/colormap.py in _glsl_mix(controls=array([0.1, 0.9], dtype=float32), colors=array([[0., 0., 0., 1.],
    143 
    144     """
--> 145     assert (controls[0], controls[-1]) == (0., 1.)
```

## Expected behavior

If Colormap is properly constructed (and `colorbar` works) then it should not raise an exception when applying to layer. 

## Possible solutions

1. Fix controls and colors before passing to vispy (napari/napari/_vispy/vispy_image_layer.py line 128) by adding a copy of first and last color with respectively coordinate 0 and 1 if absent.
2. Raise an exception if trying to create a colormap with first and last coordinate different to (0, 1)

I prefer the first solution.

## Environment
last master (42d48e3d63aa850ec18a3a6167e7441275c94303)

### PR DESCRIPTION
**Pull request by Czaki**
# Description

Fix #2400

## Type of change

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by sofroniewn** [comment]
I don't like adding the `vispy_controls` attribute. I'd rather leave our colormap object simpler, and more independent of vispy. 

The more I look at it the more I find it magically to be guessing that a user wanted us to add `0` and `1` to the controls, I think instead we should raise an error if controls don't being with `0`, end with `1`, and be strictly increasing. We could check for all three properties in the validator here https://github.com/napari/napari/blob/1dbb8bfbd07e9aae1fbb0e0ac864abf88e0d8af4/napari/utils/colormaps/colormap.py#L60

We should then add a line to the documentation saying controls must start with `0` and `1` and be strictly increasing.

Maybe we can ask @jni for a third opinion if you disagree, but to me that is the cleanest and most robust solution.

**Comment by Czaki** [comment]
> I don't like adding the `vispy_controls` attribute. I'd rather leave our colormap object simpler, and more independent of vispy.

How it introduces a dependency on vispy? Its only transforms inner representation to this one which is accepted by vispy. Maybe this method should be private?

**Comment by sofroniewn** [comment]
> How it introduces a dependency on vispy? Its only transforms inner representation to this one which is accepted by vispy. Maybe this method should be private?

It's not a dependency, it's just a coupling that adds complexity. My preference is still to just check the input and raise an error if the input does not conform to what it needs to be

**Comment by Czaki** [comment]
For me forcing 0, 1 will force to add additional complexity for colormap create widget. But it is doable.

**Comment by jni** [comment]
I think I see both points, the behaviour is magical but convenient. In the past I have often regretted magical behaviour, as it can lead to difficult to debug edge cases. But I'm ok including it here for now. At any rate, I see @sofroniewn's concern that these changes would make our colormaps behave differently in the context of vispy vs outside of it, so I support his idea of adding the 0-1 bounds for all colormaps, not just for vispy, and doing this in the validator method.

**Comment by Czaki** [comment]
Why behave differently? This implementation mimics current color bar behavior.

**Comment by jni** [comment]
@Czaki because self.map will use self.controls, not self.vispy_controls. Using the validator methods to pad the control points will achieve what you want without doing anything related to vispy, won't it?

**Comment by Czaki** [comment]
But `vispy_controls` is the only minimum set of points to restore colormap on the range `[0,1]`.

**Comment by jni** [comment]
sorry, I don't think we are understanding each other, but I might be missing the problem... I'm saying, yes, vispy_controls is what you need to get [0, 1]. Let's make *controls* do the same thing and then we don't need vispy_controls: controls should provide the colormap in [0, 1].

**Comment by Czaki** [comment]
> sorry, I don't think we are understanding each other, but I might be missing the problem... I'm saying, yes, vispy_controls is what you need to get [0, 1]. Let's make _controls_ do the same thing and then we don't need vispy_controls: controls should provide the colormap in [0, 1].

You think about autofix or raise an exception?

**Comment by sofroniewn** [comment]
> You think about autofix or raise an exception?

I think @jni is proposing to do the autofix. Let's just do the autofix inside the validator, and adjust the doc string as such. You might as well also check that the control points are strictly increasing after the fix, and raise an error if not.

**Comment by Czaki** [comment]
> I think @jni is proposing to do the autofix. Let's just do the autofix inside the validator, and adjust the doc string as such. You might as well also check that the control points are strictly increasing after the fix, and raise an error if not.

Could I manipulate one variable when validate second one? (when add controls the need to add colors)

**Comment by sofroniewn** [comment]
> Could I manipulate one variable when validate second one? (when add controls the need to add colors)

Yes, you can use a [root validator](https://pydantic-docs.helpmanual.io/usage/validators/#root-validators) if needed. You can also use that to check the length of the colors and the control points remain in sync

**Comment by Czaki** [comment]
Did I should add a warning message?

**Comment by sofroniewn** [comment]
> Did I should add a warning message?

Yes, let's add warnings. Can you then test for them in your test too

File: napari/_tests/test_examples.py [hunk 1]
```diff
@@ -44,6 +44,8 @@ def qapp():
 @pytest.mark.parametrize("fname", examples)
 def test_examples(qapp, fname, monkeypatch, capsys):
     """Test that all of our examples are still working without warnings."""
+    if fname == "embed_ipython.py":
```
Comment (sofroniewn) [inline napari/_tests/test_examples.py]: Do we need this? Can it come in it's own PR - it seem unrelated

File: napari/utils/colormaps/colormap.py [hunk 2]
```diff
@@ -63,8 +63,24 @@ def _check_controls(cls, v, values):
...
             return np.linspace(0, 1, n_controls)
+        if not np.array_equal(v, sorted(v)):
+            raise ValueError("Coords needs to be sorted in ascending order")
         return v
 
+    @root_validator
+    def check_bound_coords(cls, values):
+        colors = values['colors']
+        controls = values['controls']
+        if controls[0] != 0:
```
Comment (sofroniewn) [inline napari/utils/colormaps/colormap.py:75]: Warnings here seem good

File: napari/utils/colormaps/colormap.py [hunk 3]
```diff
@@ -1,7 +1,7 @@
 from enum import Enum
 
 import numpy as np
-from pydantic import validator
+from pydantic import root_validator, validator
 
 from ..events import EventedModel
 from ..events.custom_types import Array
```
Comment (sofroniewn) [inline napari/utils/colormaps/colormap.py:8]: Can you add to the doc string to explain the new behaviour where we add coordinates

File: napari/utils/colormaps/colormap.py [hunk 4]
```diff
@@ -63,8 +63,24 @@ def _check_controls(cls, v, values):
...
+        colors = values['colors']
+        controls = values['controls']
+        if controls[0] != 0:
+            controls = np.concatenate(([0], controls))
+            colors = np.concatenate(([colors[0]], colors))
+        if controls[-1] != 1:
+            controls = np.concatenate((controls, [1]))
+            colors = np.concatenate((colors, [colors[-1]]))
+        values['colors'] = colors
+        values['controls'] = controls
```
Comment (sofroniewn) [inline napari/utils/colormaps/colormap.py:99]: Can you add an additional check here that both of these have the same length and raise an error if not

File: napari/_tests/test_examples.py (hunk 1 above)
Comment (Czaki) [reply napari/_tests/test_examples.py]: This is to prevent test fail when locally run of pytest. I could remove it when PR is ready.

Comment (tlambert03) [reply napari/_tests/test_examples.py]: This should have been skipped already by line 16 above... is it not skipping for you?  have you merged master?

File: napari/utils/colormaps/colormap.py (hunk 2 above)
Comment (sofroniewn) [reply napari/utils/colormaps/colormap.py:75]: You'll want to catch it with pytest though as it is expected

**Comment by Czaki** [comment]
It looks like I do not understand all usage of Colormaps.
There is a failure of `test_empty_layer_with_edge_colormap` which uses a single point `Colormap`.

Could any of you check why using single-point colormap is expected and how should be handled?

**Comment by sofroniewn** [comment]
I think there might be something funky happening here with the `grays` colormap. If you change it to `gray` does the test now pass? I've noticed weird things with that one in the past too. It could be that vispy has a different default way to magically use a single point colormap. Maybe a reason to go back to just raising an error (which could then exclude the bad `grays` map). I'm not sure though, I can try looking in more when I get the chance

File: napari/_tests/test_examples.py (hunk 1 above)
Comment (Czaki) [reply napari/_tests/test_examples.py]: It looks like it was solved after I start this PR. But now my change is obsolete.

**Comment by Czaki** [comment]
This is not from vispy, but from napari and I think that it should be handled on napari side:

https://github.com/napari/napari/blob/f801f184d2f70ddf90e20c5b13fff9f32f9fb18f/napari/utils/colormaps/colormap_utils.py#L85-L90

I push my suggestion of solve to this PR.

File: napari/utils/colormaps/colormap_utils.py [hunk 5]
```diff
@@ -86,8 +86,10 @@ def convert_vispy_colormap(colormap, name='vispy'):
     # but if they do, we want to use it
     if hasattr(colormap, '_controls'):
         controls = colormap._controls
+        colors = colormap.colors.rgba
     else:
-        controls = np.zeros((0,))
+        controls = np.linspace(0, 1)
+        colors = np.array([colormap.map(x) for x in controls])
```
Comment (Czaki) [inline napari/utils/colormaps/colormap_utils.py:92]: In this place, interpolation is for 50 points. If I good read vispy code all problematic colormaps are linear so it will be enough to use 2 points. But approaches used in vispy code allow implementing any nonlinear mapping. So I decide to use 50 points approximation.
